services/wallet/saved_addresses.go

```diff
@@ -15,41 +19,169 @@ type SavedAddress struct {
+func (sam *SavedAddressesManager) GetSavedAddressesForChainID(chainID uint64) ([]SavedAddress, error) {
+	rows, err := sam.db.Query(fmt.Sprintf("SELECT %s FROM saved_addresses WHERE network_id = ? AND removed != 1", rawQueryColumnsOrder), chainID)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	addresses, _, err := getRawSavedAddressesFromDBRows(rows)
+	return addresses, err
+}
+
+func (sam *SavedAddressesManager) GetSavedAddresses() ([]SavedAddress, error) {
+	rows, err := sam.db.Query(fmt.Sprintf("SELECT %s FROM saved_addresses WHERE removed != 1", rawQueryColumnsOrder))
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	addresses, _, err := getRawSavedAddressesFromDBRows(rows)
+	return addresses, err
+}
+
+// Provide access to the soft-delete and sync metadata
+func (sam *SavedAddressesManager) GetRawSavedAddresses() ([]SavedAddress, []SavedAddressMeta, error) {
+	rows, err := sam.db.Query(fmt.Sprintf("SELECT %s FROM saved_addresses", rawQueryColumnsOrder))
+	if err != nil {
+		return nil, nil, err
+	}
+	defer rows.Close()
+
+	return getRawSavedAddressesFromDBRows(rows)
+}
+
+func (sam *SavedAddressesManager) addRawSavedAddress(sa SavedAddress, meta SavedAddressMeta) error {
+	insert, err := sam.db.Prepare("INSERT OR REPLACE INTO saved_addresses (network_id, address, name, favourite, removed, sync_clock, update_clock) VALUES (?, ?, ?, ?, ?, ?, ?)")
 	if err != nil {
 		return err
 	}
-	_, err = insert.Exec(sa.ChainID, sa.Address, sa.Name, sa.Favourite)
+	_, err = insert.Exec(sa.ChainID, sa.Address, sa.Name, sa.Favourite, meta.Removed, meta.SyncClock, meta.UpdateClock)
 	return err
 }
 
-func (sam *SavedAddressesManager) DeleteSavedAddress(chainID uint64, address common.Address) error {
-	_, err := sam.db.Exec(`DELETE FROM saved_addresses WHERE address = ? AND network_id = ?`, address, chainID)
+func (sam *SavedAddressesManager) AddOrUpdateSavedAddress(sa SavedAddress) (updatedClock int64, err error) {
+	updateClock := time.Now().Unix()
+	err = sam.addRawSavedAddress(sa, SavedAddressMeta{false, syncClockCreatedEditedHere, sql.NullInt64{Int64: updateClock, Valid: true}})
+	if err != nil {
+		return 0, err
+	}
+	return updateClock, nil
+}
+
+func (sam *SavedAddressesManager) checkIfNewerChange(ChainID uint64, Address common.Address, updateClock int64) (newer bool, err error) {
+	rows, err := sam.db.Query("SELECT update_clock FROM saved_addresses WHERE network_id = ? AND address = ?", ChainID, Address)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		var dbUpdateClock int64
+		err := rows.Scan(&dbUpdateClock)
+		if err != nil {
+			return false, err
+		}
+
+		return dbUpdateClock <= updateClock, nil
+	}
+	return true, nil
+}
+
+func (sam *SavedAddressesManager) AddSavedAddressIfNewerUpdate(sa SavedAddress, syncClock int64, updateClock int64) (insertedOrUpdated bool, err error) {
+	newer, err := sam.checkIfNewerChange(sa.ChainID, sa.Address, updateClock)
```

Comment:
I would have a the two things in a transaction, you can check if newer and update if so in a single go